### PR TITLE
a packet of small optimizations

### PR DESCRIPTION
**Pull request by pennae**
# Motivation

'tis the season for making nix faster again. this series gives anywhere from 7-10% eval performance improvement in our tests (mostly just building our system config).

# Context

none of these are particularly invasive. each patch can stand on its own too, should the entire series be a bit too much.

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by fricklerhandwerk** [comment]
@roberth this is definitely worth a release note.

**Comment by fricklerhandwerk** [comment]
A 7-10% performance improvement is definitely worth talking about.

File: src/libexpr/eval-inline.hh [hunk 1]
```diff
@@ -104,11 +104,16 @@ void EvalState::forceValue(Value & v, Callable getPos)
...
+        } catch (InfiniteRecursionError & e) {
+            // only one black hole can *throw* in any given eval stack so we need not
+            // check whether the position is set already.
+            if (v.isBlackhole())
+                e.err.errPos = positions[getPos()];
+            throw;
+        }
     }
-    else if (v.isBlackhole())
-        error("infinite recursion encountered").atPos(getPos()).template debugThrow<EvalError>();
```
Comment (roberth) [inline src/libexpr/eval-inline.hh:111]: I'll rebase https://github.com/NixOS/nix/pull/8623 after this :smiling_face_with_tear:

File: src/libexpr/eval.cc [hunk 2]
```diff
@@ -1666,18 +1667,19 @@ void EvalState::callFunction(Value & fun, size_t nrArgs, Value * * args, Value &
                 for (size_t i = 0; i < argsLeft; ++i)
                     vArgs[argsDone + i] = args[i];
 
-                auto name = primOp->primOp->name;
+                auto fn = primOp->primOp;
                 nrPrimOpCalls++;
-                if (countCalls) primOpCalls[name]++;
+                if (countCalls && !fn->hideInDiagnostics) primOpCalls[fn->name]++;
```
Comment (roberth) [inline src/libexpr/eval.cc:1672]: Technically a primop call, so that should be ok.

```suggestion
                // This will count black holes, but that's ok, because unrecoverable errors are rare.
                if (countCalls) primOpCalls[fn->name]++;
```

File: src/libexpr/eval.cc [hunk 3]
```diff
@@ -1621,15 +1621,16 @@ void EvalState::callFunction(Value & fun, size_t nrArgs, Value * * args, Value &
                 return;
             } else {
                 /* We have all the arguments, so call the primop. */
-                auto name = vCur.primOp->name;
+                auto fn = vCur.primOp;
 
                 nrPrimOpCalls++;
-                if (countCalls) primOpCalls[name]++;
+                if (countCalls && !fn->hideInDiagnostics) primOpCalls[fn->name]++;
```
Comment (roberth) [inline src/libexpr/eval.cc:1627]: Technically a primop call, so that should be ok.

```suggestion
                // This will count black holes, but that's ok, because unrecoverable errors are rare.
                if (countCalls) primOpCalls[fn->name]++;
```

File: src/libexpr/eval.hh [hunk 4]
```diff
@@ -77,6 +77,14 @@ struct PrimOp
      */
     std::optional<ExperimentalFeature> experimentalFeature;
 
+    /**
+     * Whether to hide this primop in diagnostics.
+     *
+     * Used to hide the fact that black holes are primop applications from
+     * both stack traces and call counting.
```
Comment (roberth) [inline src/libexpr/eval.hh:84]:
```suggestion
     * stack traces.
```

File: src/libexpr/primops.cc [hunk 5]
```diff
@@ -4263,6 +4263,29 @@ static RegisterPrimOp primop_splitVersion({
 });
 
 
+static void prim_blackHoleFn(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    state.error("infinite recursion encountered")
+        .debugThrow<InfiniteRecursionError>();
+}
+
+static PrimOp primop_backHole = {
```
Comment (roberth) [inline src/libexpr/primops.cc:4272]:
```suggestion
static PrimOp primop_blackHole = {
```

File: src/libexpr/value.hh [hunk 6]
```diff
@@ -447,6 +442,19 @@ public:
...
+extern Value prim_blackHole;
+
+inline bool Value::isBlackhole() const
+{
+    return internalType == tApp && app.left == &prim_blackHole;
+}
+
+inline void Value::mkBlackhole()
+{
+    mkApp(&prim_blackHole, &prim_blackHole);
```
Comment (roberth) [inline src/libexpr/value.hh:454]: May be sufficient to only set the type and `left`, but haven't verified. Already ok as is.

File: src/libexpr/primops.cc [hunk 7]
```diff
@@ -4263,6 +4263,29 @@ static RegisterPrimOp primop_splitVersion({
...
+    .name = "«blackHole»",
+    .args = {},
+    .fun = prim_blackHoleFn,
+    .hideInDiagnostics = true,
+};
+
+static Value makeBlackHole()
+{
+    Value v;
+    v.mkPrimOp(&primop_backHole);
```
Comment (roberth) [inline src/libexpr/primops.cc:4282]:
```suggestion
    v.mkPrimOp(&primop_blackHole);
```

**Review by roberth — COMMENTED** [review]
Wonderful!

File: src/libcmd/built-path.cc [hunk 8]
```diff
@@ -12,9 +12,9 @@ namespace nix {
     bool MY_TYPE ::operator COMPARATOR (const MY_TYPE & other) const \
     { \
         const MY_TYPE* me = this; \
-        auto fields1 = std::make_tuple<const CHILD_TYPE &, const FIELD_TYPE &>(*me->drvPath, me->FIELD); \
+        auto fields1 = std::tie(*me->drvPath, me->FIELD); \
         me = &other; \
-        auto fields2 = std::make_tuple<const CHILD_TYPE &, const FIELD_TYPE &>(*me->drvPath, me->FIELD); \
+        auto fields2 = std::tie(*me->drvPath, me->FIELD); \
```
Comment (Ericson2314) [inline src/libcmd/built-path.cc:17]: We should just get rid of this header and use the new `operator <=> (const MyType &) const = default;`

File: src/libexpr/eval-inline.hh (hunk 1 above)
Comment (roberth) [reply src/libexpr/eval-inline.hh:111]: While I get the sentiment, I think #8623 will benefit from this PR by moving the only hot path addition out of the hot path.
Most of it happens in `catch` blocks that shouldn't affect performance because they don't run very often at all.

**Comment by thufschmitt** [comment]
> tl;dr nix really needs performance tracking as part of CI

:100: 

https://hydra.nixos.org/job/nix/master/metrics.nixpkgs#tabs-charts does a bit of that, but since it runs on Hydra, it only happens after the fact (and hardly anyone looks at it or even knows it exists).

Related: https://github.com/NixOS/nix/issues/4897

File: src/libexpr/eval.cc [hunk 9]
```diff
@@ -1621,15 +1621,15 @@ void EvalState::callFunction(Value & fun, size_t nrArgs, Value * * args, Value &
                 return;
             } else {
                 /* We have all the arguments, so call the primop. */
-                auto name = vCur.primOp->name;
+                auto fn = vCur.primOp;
```
Comment (roberth) [inline src/libexpr/eval.cc:1624]: IIUC:

```suggestion
                auto * fn = vCur.primOp;
```

Relevant when reviewing for performance.

**Review by roberth — APPROVED** [review]
Tiny suggestion, and LGTM.

**Comment by roberth** [comment]
> reviewer bandwidth?

Just that.

> we have half an idea of how to get rid of tApp too

Another pseudo-`Expr`? Not an obvious win because of `PseudoExprApp : Expr` (or something) allocations, but maybe?

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
Thank you so much @pennae!
